caddyhttp/httpserver/replacer.go

```diff
@@ -132,6 +147,36 @@ func NewReplacer(r *http.Request, rr *ResponseRecorder, emptyValue string) Repla
 	return rep
 }
 
+func canLogRequest(r *http.Request) (canLog bool) {
+	if r.Method == "POST" || r.Method == "PUT" {
+		for _, cType := range r.Header[headerContentType] {
+			// the cType could have charset and other info
+			if strings.Index(cType, contentTypeJSON) > -1 || strings.Index(cType, contentTypeXML) > -1 {
+				canLog = true
+				break
+			}
+		}
+	}
+	return
+}
+
+// readRequestBody reads the request body and sets a
+// new io.ReadCloser that has not yet been read.
+func readRequestBody(r *http.Request, n int64) ([]byte, error) {
+	body, err := ioutil.ReadAll(io.LimitReader(r.Body, n))
+	if err != nil {
+		return nil, err
+	}
+
+	mr := io.MultiReader(
+		bytes.NewBuffer(body),
+		r.Body,
+	)
+
+	r.Body = ioutil.NopCloser(mr)
```

Comment:
Fixed.  I decided to read the remaining data from the HttpRequest body and closing it.